### ISSUE TITLE
Spurious complaint about return value in deferred function

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#4488]
Mypy incorrectly complains about the return statement in this program (with `--strict-optional`):

```py
def f() -> str:
    y  # Causes f to be deferred
    x = None
    if int():
        x = ''
    if x is None:
        x = ''
    return x  # Incompatible return value type (got "Optional[Any]", expected "str")

y = int()
```

The expected behavior is clean output.

Note that `f` is deferred. The problem goes away when `f` is not deferred.

**Comment by JukkaL** [comment]
Slightly tweaked example illustrates another problem:

```py
def f() -> str:
    y
    x = None
    if int():
        x = ''
    if x is None:
        x = ''
        x[0]  # Value of type "None" is not indexable   <-- Unexpected
    return x  # Incompatible return value type (got "Optional[Any]", expected "str")

y = int()
```
